fixed incorrect literal_float regex

wate123 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Fixes #16407 
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Changed
`pat = r"[-+]?((\d*\.\d+)|(\d+\.?))(eE[-+]?\d+)?"`
to
`pat = r"[-+]?((\d*\.\d+)|(\d+\.?))([eE][-+]?\d+)?"`

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

wate123 committed:
fixed literal_float regex

oscarbenjamin commented:
Can you give examples of the difference? Why is this an improvement?

Can you then add those examples to the tests in some way?

wate123 commented:
> Changed
> `pat = r"[-+]?((\d*\.\d+)|(\d+\.?))(eE[-+]?\d+)?"`
> to
> `pat = r"[-+]?((\d*\.\d+)|(\d+\.?))([eE][-+]?\d+)?"`

with[eE] The regex will match e or E in the input.
such as 5e10.

oscarbenjamin commented:
Can you give an example of a string that would match one regex and not the other?

Can you give an example of how that makes a difference in practice for SymPy?

smichr commented:
It works because the float matches up to the exponential part. That being the case, it is perhaps not necessary to be perfect, just sufficient for the purpose of the Float routine. And apparently it works without that part so the simplest fix is to delete it: `r"[-+]?((\d*\.\d+)|(\d+\.?))`. If it is kept, it might be better to make it
```diff
diff --git a/sympy/core/numbers.py b/sympy/core/numbers.py
index b7c01bd..735dcc4 100644
--- a/sympy/core/numbers.py
+++ b/sympy/core/numbers.py
@@ -147,10 +147,10 @@ def _decimal_to_Rational_prec(dec):
     return rv, prec


-def _literal_float(f):
+_refloat = regex.compile(r"[-+]?((\d*\.\d+)|(\d+\.?\d?))([eE][-+]?\d+)?")
+def _literal_float(f, d=False):
     """Return True if n can be interpreted as a floating point number."""
-    pat = r"[-+]?((\d*\.\d+)|(\d+\.?))(eE[-+]?\d+)?"
-    return bool(regex.match(pat, f))
+    return bool(_refloat.fullmatch(f))

 # (a,b) -> gcd(a,b)
```

smichr commented:
Here is a test of the pattern:
```python
>>> bad = []
>>> good=[]
>>> for p in cartes('+- ','1 ','. ','1 ','eE ','+- ','1 '):
...  n = ''.join(p).replace(' ','')
...  try:
...    ok=float(n)
...    good.append(n)
...  except: bad.append(n)
...
>>> len(good)
152
>>> len(bad)
280
>>> from sympy.core.numbers import _refloat
>>> assert all(_refloat.fullmatch(i) for i in good)
>>> assert not any(_refloat.fullmatch(i) for i in bad)
```

smichr committed:
omit exponential matching in _literal_float

The Float routine will emit an error if the number is not in the right format. The trailing part hasn't been working because of the eE not being written as [eE]. It has been sufficient to see that the number starts like a floating point number so that is not being changed.

To be more strict and only match an exact float -- including the exponential part -- the following tested diff could be applied:

diff --git a/sympy/core/numbers.py b/sympy/core/numbers.py
index b7c01bd..735dcc4 100644
--- a/sympy/core/numbers.py
+++ b/sympy/core/numbers.py
@@ -147,10 +147,10 @@ def _decimal_to_Rational_prec(dec):
     return rv, prec


-def _literal_float(f):
+_refloat = regex.compile(r"[-+]?((\d*\.\d+)|(\d+\.?\d?))([eE][-+]?\d+)?")
+def _literal_float(f, d=False):
     """Return True if n can be interpreted as a floating point number."""
-    pat = r"[-+]?((\d*\.\d+)|(\d+\.?))(eE[-+]?\d+)?"
-    return bool(regex.match(pat, f))
+    return bool(_refloat.fullmatch(f))

 # (a,b) -> gcd(a,b)
 
bad = []
good=[]
for p in cartes('+- ','1 ','. ','1 ','eE ','+- ','1 '):
 n = ''.join(p).replace(' ','')
 try:
   ok=float(n)
   good.append(n)
 except: bad.append(n)
assert len(good) == 152
assert len(bad) == 280
from sympy.core.numbers import _refloat
assert all(_refloat.fullmatch(i) for i in good)
assert not any(_refloat.fullmatch(i) for i in bad)
